### PR TITLE
build: add support for typescript

### PR DESCRIPTION
**Pull request by matthewgallo**
Contributes to #4171 

This PR includes the initial changes required to support typescript in our monorepo (`ibm-products`, `ibm-products-community`, `core`). In doing so, this involved updating our babel config package (`config/babel-preset-ibm-cloud-cognitive`) which is used across other packages as a shared config, similarly there were updates required to the jest config package (`config/jest-config-ibm-cloud-cognitive`) to support files/components using typescript.

I also discovered after getting everything running via the new build script that we had some circular dependencies, those have been addressed in this PR as well with some refactoring.

I also removed the local copy of `config/typescript-config-carbon` that I had initially in this PR and was able to pull in this config from `node_modules`.

I have migrated `Saving.tsx` to make sure that everything is working as expected since this is one of our smaller components. Another way I tested these changes was by building `ibm-products` and copying the built directories into an example app's `node_modules`.

#### What did you change?
Config related files (babel, eslint, storybook, jest)

#### How did you test and verify your work?
Ran new build script and double checked built directories.

### CODE REVIEW

File: config/babel-preset-ibm-cloud-cognitive/index.js
Comment (elycheea) [inline config/babel-preset-ibm-cloud-cognitive/index.js:2]:
```suggestion
 * Copyright IBM Corp. 2020, 2024
```

Was this intentionally changed?

Comment (matthewgallo) [reply config/babel-preset-ibm-cloud-cognitive/index.js]: No, this was just accidental 😅

**Review by elycheea — DISMISSED** [review]
Quite a few circular dependencies you had to resolve 🙈
We should definitely take more note going forward.

Also tested locally with a test app and seems good to me, but probably worth waiting until after tomorrow’s release to merge anyway. :)

**Review by elycheea — APPROVED** [review]

**Pull request merged**
